### ISSUE TITLE
Gem failsafe error generating 100ks of events when no network connection

### ISSUE THREAD

**Issue by edave** [rollbar/rollbar-gem#632]
Hi There! We're running into a very annoying problem w/ Rollbar's Ruby gem on a Rails installation where Rollbar itself is generating 100k errors it then tries to report. An example of this is this item:

Title:
```
Failsafe from rollbar-gem. SocketError: "Failed to open TCP connection to api.rollbar.com:443 (getaddrinfo: nodename nor servname provided, or not known)" in /.rvm/rubies/ruby-2.3.4/lib/ruby/2.3.0/net/http.rb:882:in `rescue in block in connect

Message:
```
Failsafe from rollbar-gem. SocketError: "Failed to open TCP connection to api.rollbar.com:443 (getaddrinfo: nodename nor servname provided, or not known)" in /.rvm/rubies/ruby-2.3.4/lib/ruby/2.3.0/net/http.rb:882:in `rescue in block in connect': error in process_item
```

We suspect it's because something happens where the gem sense an internal connectivity issue, and queues an occurrence to be reported, but this exponentially spawns more errors. This may occur when the machine has a valid connection to the internet, but the internet service is poor or non-existent, resulting in DNS lookup failures, etc.

We have a Rails 4 app, Rollbar Ruby gem 2.15.0 and are using Sidekiq/Redis to schedule reporting to Rollbar in the background. We have seen this occur across multiple Ubuntu 16.04 installs and MacOS 10.12.6

**Comment by rokob** [comment]
We should probably only try to send a single failsafe error, as in if there is one in the queue when we try to add another one then we should throw the previous one out. I think we probably still want to keep these fail safe errors in there, but we definitely shouldn't send more than one of the same kind.

We could also specifically ignore SocketErrors.

As a workaround before we fix this, you could use `exception_level_filters` on the `SocketError` or by raising `Rollbar::Ignore` inside the `before_process` hook when these are identified.

**Comment by rivkahstandig3636** [comment]
Hi there, I’m closing out all issues opened before 2018 that haven’t had any activity on them since the start of this year. If this is still an issue for you, please comment here and we can reopen this. Thanks!

**Comment by fcheung** [comment]
Hi, I just ran into this in a small sinatra app with rollbar 2.16.2- gigabytes of errors of the form

```
App 25655 stderr: I, [2018-06-07T19:18:14.778658 #25898]  INFO -- : [Rollbar] Scheduling item
App 25655 stderr: I, [2018-06-07T19:18:14.778932 #25898]  INFO -- : [Rollbar] Sending item
App 25655 stderr: E, [2018-06-07T19:18:14.779713 #25898] ERROR -- : [Rollbar] Error processing the item: NoMethodError, Failed to open TCP connection to api.rollbar.com:443 (undefined method `[]' for nil:NilClass). Item: {"access_token"=>"6743e43db31c408e90b13946f5532a1b", "data"=>{:level=>"erApp 25655 stderr: E, [2018-06-07T19:18:45.727432 #25906] ERROR -- : [Rollbar] Error processing the item: NoMethodError, Failed to open TCP connection to api.rollbar.com:443 (undefined method `[]' for nil:NilClass). Item: {"access_token"=>"6743e43db31c408e90b13946f5532a1b", "data"=>{:level=>"error", :environment=>"production", :body=>{:message=>{:body=>"Failsafe from rollbar-gem. NoMethodError: \"Failed to open TCP connection to api.rollbar.com:443 (undefined method `[]' for nil:NilClass)\" in /usr/local/rbenv/versions/2.4.3/lib/ruby/2.4.0/net/http.rb:906:in `rescue in block in connect': error in process_item"}}, :notifier=>{:name=>"rollbar-gem", :version=>"2.16.2"}, :custom=>{:orig_uuid=>nil, :orig_host=>nil}, :internal=>true, :failsafe=>true}}
```

( The underlying error was caused by https://github.com/sinatra/sinatra/issues/1441 but I don't think that is super relevant)

**Comment by joesustaric** [comment]
We have experienced this issue as well.
Our app is using Sinatra v2.0.3 and rollbar gem v2.17.0

I've been able to recreate the error quite easily.. it will continue to fill the logs until our instance runs out of disk space or until the connectivity is regained with the rollbar api.

Thanks for looking at this.

**Comment by waltjones** [comment]
@joesustaric,

> I've been able to recreate the error quite easily.

Can you say more about how you are doing this? I am able to produce a single failure and log entry, but not a runaway looping or recursive condition that would fill up the log file as described.

**Comment by joesustaric** [comment]
Sure @waltjones .
If I remember correctly all I did was disable internet access from the machine running the app to simulate that the rollbar API was offline or unreachable etc.. 
Then attempt to report an error..

Then watch the log fill up with retries.

**Comment by waltjones** [comment]
@joesustaric, Interesting. That's also what I'm doing, though using latest master for rollbar-gem. Thanks for the reply.

**Comment by joesustaric** [comment]
yeah very interesting .. See if it happens with v2.17.0 maybe? validate if it's been resolved? possibly could have been a bug in a dependency 🤷‍♂️ - Thanks 👍

**Comment by johnsyweb** [comment]
I have reproduced this using v2.22.1 using the method @joesustaric described.

**Comment by johnsyweb** [comment]
I have reproduced this using v2.22.1 using the method @joesustaric described.

```
--------✁-------
E, [2019-10-30T01:18:07.645214 #26297] ERROR -- : [Rollbar] Sending failsafe response due to SocketError: "Failed to open TCP connection to api.rollbar.com:443 (getaddrinfo: Temporary failure in name resolution)" in /usr/lib/ruby/2.5.0/net/http.rb:939:in `rescue in block in connect': error in process_item
I, [2019-10-30T01:18:07.645246 #26297]  INFO -- : [Rollbar] Scheduling item
E, [2019-10-30T01:18:07.645533 #26297] ERROR -- : [Rollbar] Item: #<Rollbar::Item:0x00007f59f0338230>
I, [2019-10-30T01:18:07.645618 #26297]  INFO -- : [Rollbar] Sending item
E, [2019-10-30T01:18:07.646215 #26297] ERROR -- : [Rollbar] Error processing the item: SocketError, Failed to open TCP connection to api.rollbar.com:443 (getaddrinfo: Temporary failure in name resolution). Item: {"access_token"=>"*****", "data"=>{:level=>"error", :environment=>"staging", :body=>{:message=>{:body=>"Failsafe from rollbar-gem. SocketError: \"Failed to open TCP connection to api.rollbar.com:443 (getaddrinfo: Temporary failure in name resolution)\" in /usr/lib/ruby/2.5.0/net/http.rb:939:in `rescue in block in connect': error in process_item"}}, :notifier=>{:name=>"rollbar-gem", :version=>"2.22.1"}, :custom=>{:orig_uuid=>nil, :orig_host=>nil}, :internal=>true, :failsafe=>true}}
E, [2019-10-30T01:18:07.646256 #26297] ERROR -- : [Rollbar] Reporting internal error encountered while sending data to Rollbar.
I, [2019-10-30T01:18:07.648890 #26297]  INFO -- : [Rollbar] Sending item
E, [2019-10-30T01:18:07.649933 #26297] ERROR -- : [Rollbar] Error processing the item: SocketError, Failed to open TCP connection to api.rollbar.com:443
--------✁--------
````

**Comment by waltjones** [comment]
@johnsyweb Thank you!

**Comment by johnsyweb** [comment]
@waltjones Is there an anticipated release date for https://github.com/rollbar/rollbar-gem/pull/909?

**Comment by waltjones** [comment]
@johnsyweb Next release. Hopefully this week.

**Comment by mayordwells** [comment]
We are still having this issue in 2020. 
Our period only started 3 days ago, and we received a mail that it has been exhausted only few days later. 

Something is draining the occurrence quickly. The period starts on the 27th (a few days ago).
At 09:05 this morning I got an e-mail that we exceeded 3500 occurrences (70%) and only one minute later we reached the limit. So 1500 occurrences in a minute??

![picturemessage_2cdfbrrs qjh](https://user-images.githubusercontent.com/1370340/78012077-247bf180-734d-11ea-861b-76b9a71de9fa.png)

Since the moment our trial period finished 2 months ago, rollbar tries to consume the whole 5,000 allocated to the free account within 2-3days, that doesn't look good, because it actually makes it more difficult to convince my boss that we need to migrate to a paid subscription now. 
<img width="1391" alt="Screenshot 2020-03-31 at 12 27 07" src="https://user-images.githubusercontent.com/1370340/78011893-dff05600-734c-11ea-9897-f8548405818d.png">

Why is this still happening? this issue has been created 3 years ago and it's still not fixed.

**Comment by waltjones** [comment]
@mayordwells The (at that time) accepted fix for this was released in 2.23.0 https://github.com/rollbar/rollbar-gem/releases/tag/v2.23.0 . Can you either post a link to the item here, or send to support@rollbar.com and mention my name.

**Comment by johnsyweb** [comment]
Without having access to @mayordwells's application or Rollbar's systems, I'd posit that this is a slightly different scenario to the one that my change in #909 sought to address.

In the Sinatra application I was working on at the time, if a connection couldn't be made to api.rollbar.com:443, the gem would try to process the item _ad infinitum_ and it would not reach Rollbar but our application's error logs would grow quickly.

I _suspect_ that in @mayordwells' case, the items are being processed by a background worker (Sidekiq?) with a built-in retry mechanism, which is why the items eventually make it to Rollbar, along with the failsafe message (perhaps one per retry?) and the quota is exceeded.
